sql/driver/driver_test.go

```diff
@@ -227,3 +229,96 @@ func TestInsecure(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// Start a transaction to update a row. After updating the row and
+// before committing the transaction, start another transaction in
+// another thread to update the same row. The other transaction succeeds,
+// resulting in the first transaction failing.
+func TestOverlappingTransactions(t *testing.T) {
+	defer leaktest.AfterTest(t)
+	s, db := setup(t)
+	defer cleanup(s, db)
+
+	// Prepare DB
+	if _, err := db.Exec(`CREATE DATABASE t`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`CREATE TABLE t.kv (k CHAR PRIMARY KEY, v char)`); err != nil {
+		t.Fatal(err)
+	}
+	if _, err := db.Exec(`INSERT INTO t.kv VALUES ($1, $2)`, "a", "b"); err != nil {
+		t.Fatal(err)
+	}
+
+	// Start first transaction
+	txn, err := db.Begin()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if _, err := txn.Exec(`UPDATE t.kv SET v = $2 WHERE k IN ($1)`, "a", "c"); err != nil {
+		t.Fatal(err)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	// Run another transaction in parallel that succeeds.
+	go func() {
```

Comment:
More I think about it perhaps we do not need this test. I'm unable to come
up with a strong reason to keep it. The other test is testing all the multi
statement transaction stuff and serializability is tested elsewhere. Should
I just kill this test?

On Mon, Aug 24, 2015, 7:55 AM Peter Mattis <notifications@github.com> wrote:

> In sql/driver/driver_test.go
> <https://github.com/cockroachdb/cockroach/pull/2049#discussion_r37743152>:
>
> > +		t.Fatal(err)
> > +	}
> > +
> > +	// Start first transaction
> > +	txn, err := db.Begin()
> > +	if err != nil {
> > +		t.Fatal(err)
> > +	}
> > +	if _, err := txn.Exec(`UPDATE t.kv SET v = $2 WHERE k IN ($1)`, "a", "c"); err != nil {
> > +		t.Fatal(err)
> > +	}
> > +
> > +	var wg sync.WaitGroup
> > +	wg.Add(1)
> > +	// Run another transaction in parallel that succeeds.
> > +	go func() {
>
> Ditto. And by running it in another goroutine you haven't actually tested
> whether the transaction is running in parallel, just that it might run in
> parallel. It would be better to do something like:
>
>   txn1, err := db.Begin()
>   txn2, err := db.Begin()
>
>   // Perform some txn1 operations
>   // Perform some txn2 operations
>
>  txn1.Commit()
>  txn2.Commit()
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/2049/files#r37743152>.
>